test/test_MVU.sh

```diff
@@ -0,0 +1,222 @@
+upgrade_dir=$(short_tmpdir test_pgtap_upgrade.upgrade)
+old_dir=$(short_tmpdir test_pgtap_upgrade.old)
+new_dir=$(short_tmpdir test_pgtap_upgrade.new)
+
+# Note: if this trap fires and removes the old directories with databases still
+# running we'll get a bunch of spew on STDERR. It'd be nice to have a trap that
+# knew what databases might actually be running.
+exit_trap() {
+    # No point in stopping on error in here...
+    set +e
+
+    # Force sudo on a debian system (see below)
+    [ -z "$ctl_separator" ] || sudo=$(which sudo)
+
+    # Do not simply stick this command in the trap command; the quoting gets
+    # tricky, but the quoting is also damn critical to make sure rm -rf doesn't
+    # hose you if the temporary directory names have spaces in them!
+    $sudo rm -rf "$upgrade_dir" "$old_dir" "$new_dir"
+}
+[ -n "$keep" ] || trap exit_trap EXIT
+debug 5 "traps: $(trap -p)"
+
+export PGDATA=$old_dir
+export PGPORT=$OLD_PORT
+
+if which pg_ctlcluster > /dev/null 2>&1; then
+    # Looks like we're running in a apt / Debian / Ubuntu environment, so use their tooling
+    ctl_separator='--'
+    old_initdb="sudo pg_createcluster $OLD_VERSION test_pg_upgrade -p $OLD_PORT -d $old_dir -- -A trust"
+    new_initdb="sudo pg_createcluster $NEW_VERSION test_pg_upgrade -p $NEW_PORT -d $new_dir -- -A trust"
+    old_pg_ctl="sudo pg_ctlcluster $PGVERSION test_pg_upgrade"
+    new_pg_ctl=$old_pg_ctl
+    # See also ../pg-travis-test.sh
+    new_pg_upgrade=/usr/lib/postgresql/$PGVERSION/bin/pg_upgrade
+else
+    ctl_separator=''
+    old_initdb="$(find_at_path "$OLD_PATH" initdb) -N"
+    new_initdb="$(find_at_path "$NEW_PATH" initdb) -N"
+    # s/initdb/pg_ctl/g
+    old_pg_ctl=$(find_at_path "$OLD_PATH" pg_ctl)
+    new_pg_ctl=$(find_at_path "$NEW_PATH" pg_ctl)
+
+    new_pg_upgrade=$(find_at_path "$NEW_PATH" pg_upgrade)
+fi
+
+
+banner "Creating old version temporary installation at $PGDATA on port $PGPORT"
+$old_initdb
+modify_config
+
+banner "Starting OLD postgres via $old_pg_ctl"
+$old_pg_ctl start $ctl_separator -w # older versions don't support --wait
+
+echo "Creating database"
+createdb # Note this uses PGPORT, so no need to wrap.
+
+echo "Installing pgtap"
+# If user requested sudo then we need to use it for the install step. TODO:
+# it'd be nice to move this into the Makefile, if the PGXS make stuff allows
+# it...
+( cd $(dirname $0)/.. && $sudo make clean install )
+
+
+banner "Loading extension"
+psql -c 'CREATE EXTENSION pgtap' # Also uses PGPORT
+
+echo "Stopping OLD postgres via $old_pg_ctl"
+$old_pg_ctl stop $ctl_separator -w # older versions don't support --wait
+
+export PGDATA=$new_dir
+export PGPORT=$NEW_PORT
+
+
+banner "Creating new version temporary installation at $PGDATA on port $PGPORT"
+$new_initdb
+modify_config
+
+echo "Running pg_upgrade"
+cd $upgrade_dir
+$new_pg_upgrade -d "$old_dir" -D "$new_dir" -b "$OLD_PATH" -B "$NEW_PATH"
+
+
+# TODO: turn this stuff on. It's pointless to test via `make regress` because
```

Comment:
@theory is there already a way to test pgTap without using installcheck? It'd be good to test an installed pgTap that's gone through a pg_upgrade to make sure it still works...

Reply:
Yeah, as long as you don't need the C library (8.2 or earlier), you can run the TAP tests directly; that's what `make test` does, IIRC. Might need to make sure `pg_prove` is installed, though.